fastcrypto-vdf/src/math/crt.rs

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::math::extended_gcd::extended_euclidean_algorithm;
+use num_bigint::BigInt;
+use num_integer::Integer;
+use num_traits::{One, Signed};
+
+/// Find the unique x such that x = a mod p and x = b mod q for relatively prime p and q and 0 <= x
+/// < pq.
+pub fn solve_simple_equation(a: &BigInt, p: &BigInt, b: &BigInt, q: &BigInt) -> Option<BigInt> {
+    if !p.is_positive() || !q.is_positive() {
+        return None;
+    }
+
+    // The moduli must be relatively prime
+    let output = extended_euclidean_algorithm(p, q);
+    if !output.gcd.is_one() {
+        return None;
+    }
+
+    let a = a.mod_floor(p);
+    let b = b.mod_floor(q);
+
+    let result = a * output.y * q + b * output.x * p;
+
+    if result.is_negative() {
+        Some(result + &(p * q))
+    } else {
+        Some(result)
+    }
+}
+
+/// Find the unique x such that x = a_i mod p_i for relatively prime p_i and 0 <= x < Prod p_i.
+pub fn solve_equation(a: &[BigInt], p: &[BigInt]) -> Option<BigInt> {
+    assert_eq!(a.len(), p.len());
+    match a.len() {
+        0 => None,
+        1 => Some(a[0].clone()),
+        2 => solve_simple_equation(&a[0], &p[0], &a[1], &p[1]),
+        _ => {
+            let x = solve_simple_equation(&a[0], &p[0], &a[1], &p[1])?;
+            let y = solve_equation(&a[2..], &p[2..])?;
```

Comment:
should we limit the depth of this recursion? if eventually the input to this function is adversarially controlled, it can fill the stack, so as a safety measure we should limit it

Reply:
Added a limit of 64 inputs.